pkg/controllers/gpu/controller.go

```diff
@@ -0,0 +1,232 @@
+	var list devicepluginv1.GpuDevicePluginList
+	if err := clnt.List(ctx, &list); err != nil {
+		return 0, err
+	}
+
+	return len(list.Items), nil
+}
+
+func (c *controller) NewDaemonSet(rawObj runtime.Object) *apps.DaemonSet {
+	devicePlugin := rawObj.(*devicepluginv1.GpuDevicePlugin)
+
+	var nodeSelector map[string]string
+	dpNodeSelectorSize := len(devicePlugin.Spec.NodeSelector)
+	if dpNodeSelectorSize > 0 {
+		nodeSelector = make(map[string]string, dpNodeSelectorSize+1)
+		for k, v := range devicePlugin.Spec.NodeSelector {
+			nodeSelector[k] = v
+		}
+		nodeSelector["kubernetes.io/arch"] = "amd64"
+	} else {
+		nodeSelector = map[string]string{"kubernetes.io/arch": "amd64"}
+	}
+
+	yes := true
+	return &apps.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    devicePlugin.Namespace,
+			GenerateName: devicePlugin.Name + "-",
+			Labels: map[string]string{
+				"app": appLabel,
+			},
+		},
+		Spec: apps.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": appLabel,
+				},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": appLabel,
+					},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name: appLabel,
+							Env: []v1.EnvVar{
+								{
+									Name: "NODE_NAME",
+									ValueFrom: &v1.EnvVarSource{
+										FieldRef: &v1.ObjectFieldSelector{
+											FieldPath: "spec.nodeName",
+										},
+									},
+								},
+							},
+							Args:            getPodArgs(devicePlugin),
+							Image:           devicePlugin.Spec.Image,
+							ImagePullPolicy: "IfNotPresent",
+							SecurityContext: &v1.SecurityContext{
+								ReadOnlyRootFilesystem: &yes,
+							},
+							VolumeMounts: []v1.VolumeMount{
+								{
+									Name:      "devfs",
+									MountPath: "/dev/dri",
+									ReadOnly:  true,
+								},
+								{
+									Name:      "sysfs",
+									MountPath: "/sys/class/drm",
+									ReadOnly:  true,
+								},
+								{
+									Name:      "kubeletsockets",
+									MountPath: "/var/lib/kubelet/device-plugins",
+								},
+							},
+						},
+					},
+					NodeSelector: nodeSelector,
+					Volumes: []v1.Volume{
+						{
+							Name: "devfs",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/dev/dri",
+								},
+							},
+						},
+						{
+							Name: "sysfs",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/sys/class/drm",
+								},
+							},
+						},
+						{
+							Name: "kubeletsockets",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/var/lib/kubelet/device-plugins",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (c *controller) UpdateDaemonSet(rawObj runtime.Object, ds *apps.DaemonSet) (updated bool) {
```

Comment:
this function does not take `nodeSelector` updates into account.

Reply:
Thanks! Updated both controllers for GPU and QAT.

Reply:
works!